android/src/main/java/android/print/PdfConverter.java

```diff
@@ -0,0 +1,130 @@
+        mWebView = new WebView(mContext);
+        mWebView.setWebViewClient(new WebViewClient() {
+            @Override
+            public void onPageFinished(WebView view, String url) {
+                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT)
+                    throw new RuntimeException("call requires API level 19");
+                else {
+                    PrintDocumentAdapter documentAdapter = mWebView.createPrintDocumentAdapter();
+                    documentAdapter.onLayout(null, getPdfPrintAttrs(), null, new PrintDocumentAdapter.LayoutResultCallback() {
+                    }, null);
+                    documentAdapter.onWrite(new PageRange[]{PageRange.ALL_PAGES}, getOutputFileDescriptor(), null, new PrintDocumentAdapter.WriteResultCallback() {
+                        @Override
+                        public void onWriteFinished(PageRange[] pages) {
+                            destroy();
+                        }
+                    });
+                }
+            }
+        });
+        mWebView.loadData(mHtmlString, "text/HTML", "UTF-8");
+    }
+
+    public PrintAttributes getPdfPrintAttrs() {
+        return mPdfPrintAttrs != null ? mPdfPrintAttrs : getDefaultPrintAttrs();
+    }
+
+    public void setPdfPrintAttrs(PrintAttributes printAttrs) {
+        this.mPdfPrintAttrs = printAttrs;
+    }
+
+    public void convert(Context context, String htmlString, File file) {
+        if (context == null)
+            throw new IllegalArgumentException("context can't be null");
+        if (htmlString == null)
+            throw new IllegalArgumentException("htmlString can't be null");
+        if (file == null)
+            throw new IllegalArgumentException("file can't be null");
+
+        if (mIsCurrentlyConverting)
+            return;
+
+        mContext = context;
+        mHtmlString = htmlString;
+        mPdfFile = file;
+        mIsCurrentlyConverting = true;
+        runOnUiThread(this);
+    }
+
+    private ParcelFileDescriptor getOutputFileDescriptor() {
+        try {
+            mPdfFile.createNewFile();
+            return ParcelFileDescriptor.open(mPdfFile, ParcelFileDescriptor.MODE_TRUNCATE | ParcelFileDescriptor.MODE_READ_WRITE);
+        } catch (Exception e) {
+            Log.d(TAG, "Failed to open ParcelFileDescriptor", e);
+        }
+        return null;
+    }
+
+    private PrintAttributes getDefaultPrintAttrs() {
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) return null;
+
+        return new PrintAttributes.Builder()
+                .setMediaSize(PrintAttributes.MediaSize.NA_GOVT_LETTER)
+                .setResolution(new PrintAttributes.Resolution("RESOLUTION_ID", "RESOLUTION_ID", 600, 600))
+                .setMinMargins(PrintAttributes.Margins.NO_MARGINS)
+                .build();
+
+    }
+
+    private void runOnUiThread(Runnable runnable) {
+        Handler handler = new Handler(mContext.getMainLooper());
+        handler.post(runnable);
```

Comment:
How so?

Reply:
It seems to start another thread, and this thread is not associated with a com.facebook.react.bridge.Promise object. I have actually tried this. I called RNHTMLtoPDF.convert() method with await and as soon as this method returned, I tried to view the generated PDF with react-native-pdf-view and it was empty. Then I made another experiment. This  time after calling await RNHTMLtoPDF.convert() I set up a 1 second timer, then I view the generated PDF. This time it was fine.

Reply:
BTW, I have added a base64 image to the HTML that I am trying to convert to PDF. I think this helped me to reveal the problem because I think it takes some time to render a base64 image and this latency helped me to find the await/async problem.